simplejson/_speedups.c

```diff
@@ -386,6 +386,9 @@ static int
 _is_namedtuple(PyObject *obj)
 {
     int rval = 0;
+    if (!PyTuple_Check(obj)) {
+        return 0;
+    }
```

Comment:
This would break compatibility, there is intentionally not a tuple requirement here https://github.com/simplejson/simplejson/blob/5a40364f951f85ed73d6aaf881f8ee9521a56f07/CHANGES.txt#L446-L450

Reply:
It looks like the CI must not have been working correctly when this was originally submitted (due to Travis-CI) but the test that this breaks is [test_namedtuple_dumps](https://github.com/simplejson/simplejson/pull/284/checks?check_run_id=3383385834#step:4:259)

```
  ======================================================================
  ERROR: test_namedtuple_dumps (simplejson.tests.test_namedtuple.TestNamedTuple)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.reGP4CJQiP/venv/lib/python3.6/site-packages/simplejson/tests/test_namedtuple.py", line 55, in test_namedtuple_dumps
      self.assertEqual(d, json.loads(json.dumps(v)))
    File "/tmp/tmp.reGP4CJQiP/venv/lib/python3.6/site-packages/simplejson/__init__.py", line 395, in dumps
      return _default_encoder.encode(obj)
    File "/tmp/tmp.reGP4CJQiP/venv/lib/python3.6/site-packages/simplejson/encoder.py", line 296, in encode
      chunks = self.iterencode(o, _one_shot=True)
    File "/tmp/tmp.reGP4CJQiP/venv/lib/python3.6/site-packages/simplejson/encoder.py", line 378, in iterencode
      return _iterencode(o, 0)
    File "/tmp/tmp.reGP4CJQiP/venv/lib/python3.6/site-packages/simplejson/encoder.py", line 273, in default
      o.__class__.__name__)
  TypeError: Object of type DuckValue is not JSON serializable
```

Reply:
I believe the reason this was happening is that a blind Mock has an `_asdict()` method so the mock was being considered to be a namedtuple (wrong - but given your logic around _asdict maybe it is what you want to happen).  Calling that method then returns another Mock (not a dict).  Which tripped up the code in `encoder_listencode_dict`.

Your suggested refactorings make sense to me.  changes pushed. PTAL.